concurrency/DdbTableManager.java

```diff
@@ -0,0 +1,206 @@
+import com.google.inject.name.Named;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class manages DynamoDb tables
+ */
+@Singleton
+@Slf4j
+public class DdbTableManager {
+
+    private final AmazonDynamoDB ddb;
+    private final Long throughput;
+    private final Sleeper sleeper;
+
+    @Inject
+    public DdbTableManager(AmazonDynamoDB ddb, @Named("rogueone.datastore.throughput") Long throughput, Sleeper sleeper) {
+        this.ddb = ddb;
+        this.throughput = throughput;
+        this.sleeper = sleeper;
+    }
+
+    /**
+     * Creates the provided app table name in DynamoDb.
+     * <p>
+     * Note that if the table name ends with "_Log" then "timestamp" and "ttl" attributes will be added automatically.
+     * <p>
+     * TODO: This method should not rely on the table naming convention. Consider two methods. One for "normal", one for "logging" tables.
+     *
+     * @return {@code true} if table was created, {@code false} otherwise
+     */
+    public boolean createAppTable(String appTableName) {
+        List<AttributeDefinition> attributeDefinitions = new ArrayList<>();
+        AttributeDefinition attributeDefinition = new AttributeDefinition("stepId", "S");
+        attributeDefinitions.add(attributeDefinition);
+
+        KeySchemaElement keySchemaElement = new KeySchemaElement("stepId", "HASH");
+        List<KeySchemaElement> keySchema = new ArrayList<>();
+        keySchema.add(keySchemaElement);
+        // add special ttl sort element, and setup the TTL attribute.
+        if (appTableName.endsWith("_Log")) {
+            keySchema.add(new KeySchemaElement().withAttributeName("timestamp").withKeyType("RANGE"));
+            attributeDefinitions.add(new AttributeDefinition().withAttributeName("timestamp").withAttributeType("N"));
+        }
+        ProvisionedThroughput provisionedThroughput = new ProvisionedThroughput(this.throughput, this.throughput);
+
+        CreateTableRequest createTableRequest = new CreateTableRequest(attributeDefinitions, appTableName, keySchema,
+            provisionedThroughput);
+
+        String tableStatus = "";
+        try {
+            tableStatus = ddb.describeTable(appTableName).getTable().getTableStatus();
+
+            // if the above statement successed, then the table already exists and cannot be recreated.
+            // This case should not happen, so returning false to indicate that something is not right.
+            log.info("Table: %s already exists", appTableName);
+            return false;
+        } catch (ResourceNotFoundException e) {
+            // Table does not exist, which is expected, so will proceeed to create it
+        }
+
+        try {
+            boolean tableMissing = true;
+            ddb.createTable(createTableRequest);
+
+            while (tableMissing) {
+                try {
+                    tableStatus = ddb.describeTable(appTableName).getTable().getTableStatus();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code appears to be waiting for a resource before it runs. You could use the waiters feature to help improve efficiency. Consider using TableExists or TableNotExists. For more information, see [https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/](https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/)